src/code/z_viscvg.c

```diff
@@ -1,70 +1,103 @@
-// alphabetically>
+#include "global.h"
 
-Gfx D_8012AC00[] = {
+/**
+ * Draws only coverage: does not retain any of the original pixel RGB, primColor is used as background colour.
+ */
+Gfx sCoverageOnlyDL[] = {
     gsDPSetOtherMode(G_AD_PATTERN | G_CD_MAGICSQ | G_CK_NONE | G_TC_CONV | G_TF_POINT | G_TT_NONE | G_TL_TILE |
                          G_TD_CLAMP | G_TP_NONE | G_CYC_1CYCLE | G_PM_NPRIMITIVE,
                      G_AC_NONE | G_ZS_PRIM | G_RM_VISCVG | G_RM_VISCVG2),
+    // (blendColor) * (cvg)
     gsDPFillRectangle(0, 0, SCREEN_WIDTH - 1, SCREEN_HEIGHT - 1),
     gsDPPipeSync(),
     gsDPSetBlendColor(0, 0, 0, 8),
     gsSPEndDisplayList(),
 };
 
-Gfx D_8012AC28[] = {
+/**
+ * Draws fog + coverage * RGB of pixels
+ *
+ * @bug This easily overflows the blender because the fog value is added to the coverage value.
+ */
+Gfx sCoverageRGBFogDL[] = {
     gsDPSetOtherMode(G_AD_PATTERN | G_CD_MAGICSQ | G_CK_NONE | G_TC_CONV | G_TF_POINT | G_TT_NONE | G_TL_TILE |
                          G_TD_CLAMP | G_TP_NONE | G_CYC_1CYCLE | G_PM_NPRIMITIVE,
                      G_AC_NONE | G_ZS_PRIM | IM_RD | CVG_DST_CLAMP | ZMODE_OPA | FORCE_BL |
                          GBL_c1(G_BL_CLR_FOG, G_BL_A_FOG, G_BL_CLR_MEM, G_BL_A_MEM) |
                          GBL_c2(G_BL_CLR_FOG, G_BL_A_FOG, G_BL_CLR_MEM, G_BL_A_MEM)),
+    // (fog RGB) * (fog alpha) + (pixel RGB) * (cvg)
     gsDPFillRectangle(0, 0, SCREEN_WIDTH - 1, SCREEN_HEIGHT - 1),
     gsSPEndDisplayList(),
 };
 
-Gfx D_8012AC40[] = {
+/**
+ * Draws coverage and RGB of pixels
+ */
+Gfx sCoverageRGBDL[] = {
     gsDPSetOtherMode(G_AD_PATTERN | G_CD_MAGICSQ | G_CK_NONE | G_TC_CONV | G_TF_POINT | G_TT_NONE | G_TL_TILE |
                          G_TD_CLAMP | G_TP_NONE | G_CYC_1CYCLE | G_PM_NPRIMITIVE,
                      G_AC_NONE | G_ZS_PRIM | IM_RD | CVG_DST_CLAMP | ZMODE_OPA | FORCE_BL |
                          GBL_c1(G_BL_CLR_IN, G_BL_0, G_BL_CLR_MEM, G_BL_A_MEM) |
                          GBL_c2(G_BL_CLR_IN, G_BL_0, G_BL_CLR_MEM, G_BL_A_MEM)),
-
+    // (pixel RGB) * (cvg)
     gsDPFillRectangle(0, 0, SCREEN_WIDTH - 1, SCREEN_HEIGHT - 1),
     gsSPEndDisplayList(),
 };
 
-Gfx D_8012AC58[] = {
+/**
+ * Two stage filtering:
+ *
+ * - Transparently blends primColor with original frame, creating uniform fog. The "cloud surface" RenderMode is used to
+ * preserve the coverage for the second stage.
+ * - Second half is the same as sCoverageRGBDL's, i.e. coverage and RGB of pixels
+ */
+Gfx sCoverageRGBPrimFogDL[] = {
     gsDPSetCombineMode(G_CC_PRIMITIVE, G_CC_PRIMITIVE),
     gsDPSetOtherMode(G_AD_NOTPATTERN | G_CD_DISABLE | G_CK_NONE | G_TC_CONV | G_TF_POINT | G_TT_NONE | G_TL_TILE |
                          G_TD_CLAMP | G_TP_NONE | G_CYC_1CYCLE | G_PM_NPRIMITIVE,
                      G_AC_NONE | G_ZS_PRIM | G_RM_CLD_SURF | G_RM_CLD_SURF2),
+    // (fog RGB) * (cvg) + (pixel RGB) * (1 - cvg).
```

Comment:
`RM_CLD_SURF` is `G_BL_CLR_IN, G_BL_A_IN, G_BL_CLR_MEM, G_BL_1MA`,

notice `p` is `G_BL_CLR_IN`, that's the combined rgb
also notice `a` is `G_BL_A_IN`, that's the combined alpha
(i.e. for both prim rgb / alpha because `gsDPSetCombineMode(G_CC_PRIMITIVE, G_CC_PRIMITIVE)` above)

basically this interpolates between the current frame buffer and a solid prim color, with prim alpha as the ratio

I guess I also wouldn't use "fog" to describe this as "fog" is something very specific in N64 graphics and this is unrelated

Reply:
```c

/**
 * Two stage filtering:
 *
 * 1. Apply a uniform colour filter by transparently blending primColor with original frame. The "cloud surface"
 * RenderMode is used to preserve the coverage for the second stage.
 * 2. Second half is the same as sCoverageRGBDL's, i.e. previous colour * cvg
 */
Gfx sCoverageRGBPrimFogDL[] = {
    gsDPSetCombineMode(G_CC_PRIMITIVE, G_CC_PRIMITIVE),
    gsDPSetOtherMode(G_AD_NOTPATTERN | G_CD_DISABLE | G_CK_NONE | G_TC_CONV | G_TF_POINT | G_TT_NONE | G_TL_TILE |
                         G_TD_CLAMP | G_TP_NONE | G_CYC_1CYCLE | G_PM_NPRIMITIVE,
                     G_AC_NONE | G_ZS_PRIM | G_RM_CLD_SURF | G_RM_CLD_SURF2),
    // stage 1 color = (primColor RGB) * (cvg) + (FB RGB) * (1 - cvg)
    gsDPFillRectangle(0, 0, SCREEN_WIDTH - 1, SCREEN_HEIGHT - 1),

    gsDPSetOtherMode(G_AD_PATTERN | G_CD_MAGICSQ | G_CK_NONE | G_TC_CONV | G_TF_POINT | G_TT_NONE | G_TL_TILE |
                         G_TD_CLAMP | G_TP_NONE | G_CYC_1CYCLE | G_PM_NPRIMITIVE,
                     G_AC_NONE | G_ZS_PRIM | IM_RD | CVG_DST_CLAMP | ZMODE_OPA | FORCE_BL |
                         GBL_c1(G_BL_CLR_IN, G_BL_0, G_BL_CLR_MEM, G_BL_A_MEM) |
                         GBL_c2(G_BL_CLR_IN, G_BL_0, G_BL_CLR_MEM, G_BL_A_MEM)),
    // final colour = (stage 1's RGB) * (cvg)
    gsDPFillRectangle(0, 0, SCREEN_WIDTH - 1, SCREEN_HEIGHT - 1),
    gsSPEndDisplayList(),
};
```
Something like this?

Reply:
I did think about something like `current` or `new` or `previous` but don't really want to get into temporal things since it's a single frame.

Reply:
Not sure what you refer to with current/new/prev? The text you wrote does use "previous"

"Stage 1" doesn't use coverage, only combined alpha (i.e. prim alpha)

Reply:
Yeah, it's tricky to describe the output of the first blend. I've had another go, see what you think.